platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java

```diff
@@ -0,0 +1,272 @@
+        }
+        g.setColor(fgColor);
+        if (buttonId == TabControlButton.ID_CLOSE_BUTTON) {
+            // Draw an "X".
+            // Use a slightly heavier line when there's a non-light background.
+            double strokeWidth = (bgColor.getAlpha() > 0 ? 1.0 : 0.8) * scaling;
+            if (scaling > 1.0) {
+                // Use a heavier line when we have more pixels available.
+                strokeWidth *= 1.5f;
+            } else if (strokeWidth < 1.0) {
+                strokeWidth = 1.0;
+            }
+            // Middle x and y.
+            double mx = width / 2.0;
+            double my = height / 2.0;
+            // Radius of the cross ("X") symbol.
+            double cr = 3.25 * scaling;
+            /* Draw the "X". Fill the Shape of the entire cross rather than painting each line
+            separately as a Stroke, to avoid the intersecting area getting a higher opacity. */
+            Stroke stroke = new BasicStroke(
+                    (float) strokeWidth, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND);
+            Area area = new Area();
+            area.add(new Area(stroke.createStrokedShape(
+                    new Line2D.Double(mx - cr, my - cr, mx + cr, my + cr))));
+            area.add(new Area(stroke.createStrokedShape(
+                    new Line2D.Double(mx + cr, my - cr, mx - cr, my + cr))));
+            g.fill(area);
+        } else if (buttonId == TabControlButton.ID_PIN_BUTTON ||
+                buttonId == TabControlButton.ID_RESTORE_GROUP_BUTTON ||
+                buttonId == TabControlButton.ID_RESTORE_BUTTON)
+        {
+            // Draw one little window on top of another.
+            int wh = round(8 * scaling);
+            // Upper right-hand corner.
+            int win1X = round(5 * scaling);
+            int win1Y = round(3 * scaling);
+            // Lower left-hand corner.
+            int win2X = round(3 * scaling);
+            int win2Y = round(5 * scaling);
+            Area win1 = getWindowSymbol(scaling, win1X, win1Y, wh, wh);
+            Area win2 = getWindowSymbol(scaling, win2X, win2Y, wh, wh);
+            // Make window 2 appear "on top of" window 1.
+            win1.subtract(new Area(win2.getBounds2D()));
+            g.fill(win1);
+            g.fill(win2);
+        } else if (buttonId == TabControlButton.ID_MAXIMIZE_BUTTON) {
+            int xy = round(3 * scaling);
+            int wh = round(10 * scaling);
+            /* Draw one larger window. The getWindowSymbol method ensures we are using the same
+            window border thickness as for ID_RESTORE_BUTTON. */
+            g.fill(getWindowSymbol(scaling, xy, xy, wh, wh));
+        } else if (buttonId == TabControlButton.ID_SLIDE_GROUP_BUTTON) {
+            // Draw a simple bar towards the bottom of the icon.
+            int barX = round(3 * scaling);
+            int barY = round(7 * scaling);
+            int barWidth = round(10 * scaling);
```

Comment:
With the VectorIcon interface, it's best to use the supplied device pixel dimensions rather than assuming the logical width you set in the constructor, i.e. do "barWidth = width - 2 * barX" instead. Otherwise you might technically draw outside the available area (due to rounding errors at non-integral HiDPI scaling factors).

(Not terribly likely to cause problems, though.)